Add ClearML Logger

thepycoder opened the pull request:
Thanks a lot for all the great work you all do :)

This PR is to add a logger for [ClearML](https://github.com/allegroai/clearml).

## Some open questions:

### The PR is only tested on [this sample](https://github.com/explosion/projects/tree/v3/pipelines/tagger_parser_ud) project so far. 
I'm a spaCy newbie, so not sure if I should do more testing :)  I got the tip to clone and adapt [this project](https://github.com/explosion/projects/tree/v3/integrations/wandb) which I'll try to do if I can get the time for it.

### Metrics visualisation
Currently, I've grouped all metrics that are more than 1 layer deep on the same graph in the webUI. The reason being that giving every one of them their own separate graph, makes the UI very full of graphs. Now, there are 2 graphs that are chock-full of metrics, but they do belong near each other and it's easy to filter them within the graph itself. Personally, I think this makes sense, but I'm all for changing it like it is in the other loggers if you prefer it that way :)

Here is a gif of what the result looks like:
![graphs_spacy](https://user-images.githubusercontent.com/11781950/194252554-dbec0582-d1e8-4138-aef4-6a491112a47f.gif)

rmitsch commented on README.md:
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
+- The full spaCy config file contents
+- Code information such as git repository, commit ID and uncommitted changes.
+- Full console output
+- Miscellanious info such as time, python version and hardware information
```
```suggestion
- Miscellaneous info such as time, python version and hardware information
```

rmitsch commented on README.md:
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
+using the [ClearML](https://www.clear.ml/) tool. To use
+this logger, ClearML should be installed and you should have initialized (using the command above).
+The logger will send all the gathered information to your ClearML server, either [the hosted free tier](https://app.clear.ml) 
+or the open source [self-hosted server](https://github.com/allegroai/clearml-server). This logger captures all the following information (all visible in the webUI):
```
```suggestion
or the open source [self-hosted server](https://github.com/allegroai/clearml-server). This logger captures all the following information (all visible in the web UI):
```

rmitsch commented on README.md:
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
+- Miscellanious info such as time, python version and hardware information
+- Output scalars:
+    - The final score is logged under the scalar `score`.
+    - Individual component scores are grouped together on 1 scalar plot, use the webUI to filter.
```
```suggestion
    - Individual component scores are grouped together on one scalar plot, use the web UI to filter.
```

rmitsch commented on spacy_loggers/clearml.py:
```diff
@@ -0,0 +1,124 @@
+        from clearml import Task, Dataset, OutputModel  # noqa: F401
+    except ImportError as exc:
+        raise ImportError(
+            "The 'mlflow' library could not be found - did you install it? "
```
```suggestion
            "The 'clearml' library could not be found - did you install it? "
```

rmitsch commented on spacy_loggers/clearml.py:
```diff
@@ -0,0 +1,124 @@
+                            )
+
+        def finalize() -> None:
+
```
```suggestion
```

rmitsch commented on spacy_loggers/clearml.py:
```diff
@@ -0,0 +1,124 @@
+
+from spacy import util
+from spacy.training.loggers import console_logger
+
```
```suggestion


```

rmitsch reviewed (changes requested):
@thepycoder Thanks a lot for your contribution, looks good! The setup with ClearML works as in the readme described. Only a few nitpicks from my side. Could you be so kind and also run ` black` on your code?

Once all comments have been addressed I'll wait a few days before my approval - to see if any of the other spaCy maintainers want to review this as well.

thepycoder commented:
Hi @rmitsch !

Thank you very much for the review! I just ran black on my changes and committed the result.

I also used the built-in "add suggestion to batch" functionality of github for the first time to accept your suggestions, it aparently resolved all the issues automatically, let me know if that's not what you wanted :) All comments were completely valid though, so no counterpoints on my end.

Let me know if you want me to do anything else!

shadeMe commented on README.md:
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
+The logger will send all the gathered information to your ClearML server, either [the hosted free tier](https://app.clear.ml) 
+or the open source [self-hosted server](https://github.com/allegroai/clearml-server). This logger captures all the following information (all visible in the web UI):
+
+- The full spaCy config file contents
```
```suggestion
- The full spaCy config file contents.
```

shadeMe commented on README.md:
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
+
+- The full spaCy config file contents
+- Code information such as git repository, commit ID and uncommitted changes.
+- Full console output
```
```suggestion
- Full console output.
```

shadeMe commented on README.md:
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
+- The full spaCy config file contents
+- Code information such as git repository, commit ID and uncommitted changes.
+- Full console output
+- Miscellaneous info such as time, python version and hardware information
```
```suggestion
- Miscellaneous info such as time, python version and hardware information.
```

shadeMe commented on README.md:
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
+- Miscellaneous info such as time, python version and hardware information
+- Output scalars:
+    - The final score is logged under the scalar `score`.
+    - Individual component scores are grouped together on one scalar plot, use the web UI to filter.
```
```suggestion
    - Individual component scores are grouped together on one scalar plot (filterable using the web UI).
```

shadeMe commented on README.md:
```diff
@@ -139,3 +140,69 @@ remove_config_values = ["paths.train", "paths.dev", "corpora.train.path", "corpo
+    - Individual component scores are grouped together on one scalar plot, use the web UI to filter.
+    - Loss values of different components are logged with the `loss_` prefix.
+
+Then optionally, depending on the setting below, the following can be captured:
```
```suggestion
In addition to the above, the following artifacts can also be optionally captured:
```

shadeMe reviewed (commented):
Looks good overall! Some minor comments below.

thepycoder commented on spacy_loggers/clearml.py:
```diff
@@ -0,0 +1,136 @@
+        config = util.dot_to_dict(config_dot)
+        task = Task.init(
+            project_name=project_name,
+            task_name=task_name if task_name else "spaCy Training",
```
Sure!

thepycoder replied:
Made it non-optional in both the function argument list and updated the readme as well.

thepycoder commented on spacy_loggers/clearml.py:
```diff
@@ -0,0 +1,136 @@
+                score = info["score"]
+                other_scores = info["other_scores"]
+                losses = info["losses"]
+                task.get_logger().report_scalar(
```
You're right. I also think indexing the dict directly was bad as well. I don't quite know what can and cannot be in the info dict, but it seems safer to use `.get()` instead, so I added that change as well.

thepycoder commented:
@shadeMe Thank you so much for the PR review.

First, I apologize, the code is ok, but it seems I was very sloppy with the docs in leaving e.g. MLFlow references in there. That should not have been on you guys to find!

Next, I implemented your suggestions, please check 2 of them that I left comments on.

Please Note: I also added an additional commit that changes the output dataset ID. It was an HTML link before, but it did not actually parse correctly on our end! So I'm reverting back to just the ID.

Thanks again for the work you guys put in here!

rmitsch commented:
Merged :tada: Thanks again, @thepycoder! And thanks to @shadeMe for doing a more thorough review than me :pray:
